components/Ripple/src/MDCStatefulRippleView.m

```diff
@@ -0,0 +1,229 @@
+        [UIColor colorWithWhite:0 alpha:kRippleDraggedAlpha];
+    _rippleColors[@(MDCRippleStateSelected | MDCRippleStateDragged)] =
+        [selectionColor colorWithAlphaComponent:kRippleDraggedAlpha];
+  }
+  self.userInteractionEnabled = YES;
+}
+
+- (UIColor *)rippleColorForState:(MDCRippleState)state {
+  UIColor *rippleColor = _rippleColors[@(state)];
+  if (rippleColor == nil && (state & MDCRippleStateDragged) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateDragged)];
+  } else if (rippleColor == nil && (state & MDCRippleStateSelected) != 0) {
+    rippleColor = _rippleColors[@(MDCRippleStateSelected)];
+  }
+
+  if (rippleColor == nil) {
+    rippleColor = _rippleColors[@(MDCRippleStateNormal)];
+  }
+  return rippleColor;
+}
+
+- (void)updateRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setRippleColor:rippleColor];
+}
+
+- (void)updateActiveRippleColor {
+  UIColor *rippleColor = [self rippleColorForState:self.state];
+  [self setActiveRippleColor:rippleColor];
+}
+
+- (void)setRippleColor:(UIColor *)rippleColor forState:(MDCRippleState)state {
+  _rippleColors[@(state)] = rippleColor;
+
+  [self updateRippleColor];
+}
+
+- (MDCRippleState)state {
+  NSInteger state = 0;
+  if (self.selected) {
+    state |= MDCRippleStateSelected;
+  }
+  if (self.rippleHighlighted) {
+    state |= MDCRippleStateHighlighted;
+  }
+  if (self.dragged) {
+    state |= MDCRippleStateDragged;
+  }
+  return state;
+}
+
+- (void)setSelected:(BOOL)selected {
+  if (!self.allowsSelection || _tapWentOutsideOfBounds) {
+    // If we disallow selection we don't want to apply any visual or state changes for selection.
+    return;
+  }
+  if (selected == _selected && self.activeRippleLayer) {
+    return;
+  }
+  _selected = selected;
+  // Go into the selected state visually.
+  if (selected) {
+    if (!self.activeRippleLayer) {
+      // If we go into the selected state manually, without coming from the highlighted state,
+      // We present the ripple overlay instantly without animation.
+      [self updateRippleColor];
+      [self beginRippleTouchDownAtPoint:_lastTouch animated:NO completion:nil];
+    } else {
+      [self updateActiveRippleColor];
+    }
+  } else {
+    // If we are no longer selecting, we cancel all the ripples.
+    [self updateRippleColor];
+    [self cancelAllRipplesAnimated:YES completion:nil];
+  }
+}
+
+- (void)setRippleHighlighted:(BOOL)rippleHighlighted {
+  if (rippleHighlighted == _rippleHighlighted) {
+    return;
+  }
+  _rippleHighlighted = rippleHighlighted;
+  // Go into the highlighted state visually.
+  if (rippleHighlighted && !_tapWentInsideOfBounds) {
+    // If ripple becomes highlighted we initiate a ripple with animation.
+    [self updateRippleColor];
+    [self beginRippleTouchDownAtPoint:_lastTouch animated:_didReceiveTouch completion:nil];
+  } else {
+    if (!self.allowsSelection && !self.selected && !self.dragged && !_tapWentOutsideOfBounds) {
```

Comment:
Should it be possible to both allow selection and have the selected state be set to YES?

Reply:
True, we don't need the !self.selected check, removed.